### DISCUSSION TITLE
DatePicker 组件，关于时区设定的问题！

### DISCUSSION THREAD

**Discussion by wowhaha7pupu** [ant-design/ant-design#50934] [Q&A] [answered]
最近专案有用到更换时区的功能，众所周知 DatePicker 组件下拉的时候会有预设圈选的时间，就是下图蓝色实线框起来的部分

![截圖 2024-09-20 16 43 06](https://github.com/user-attachments/assets/2bd33e49-cb8c-496e-9449-aa5ff9733db1)

但假如今天本机时间是20号，但如果我切换了-11的时区，这时候应该是19号，但是还是预设圈选的时间是20号
调查过一轮issue之后，发现吃到的是本机的时间！
我试过里面其他人的回复的方法：

第一种方法：使用dayjs.tz.setDefault('我的时区的时间')，这个方法没有作用，因为组件是内建使用dayjs，所以吃到的是本机的时间。

第二种方法：使用value处理，给予一个dayJs的tz处理过后的时间给组件，但最后会变成有一个选定的样式+一个预设圈选的样式，这完全是不对的。

![截圖 2024-09-20 16 27 39](https://github.com/user-attachments/assets/50c8d98c-7848-439a-81fd-4a09401a4dad)

第三种方法：使用defaultPickerValue这个官方文件有提供的属性，一样是给予一个dayJs的tz物件给它，但依旧没有效果。

以上三种方法都没有效果，甚至想要连那个预设的蓝色框框都不显示也没办法，甚至所有相关连的issue都被关闭了
真的想不到解法了才上来求救，想问一下有没有人有遇过相同的问题，然后有不错的处理方式可以提供个方向吗？
由衷感谢。

**Comment by Wxh16144** [comment, marked as answer]
花了点时间研究了一下时区这个东西，感觉还挺有意思，之前业务一直没接触过。

针对你说的蓝色框预选框（表示今日），假设现在在北京是 2024/10/1 凌晨 3点10分，如果设置了一个默认为 -5 的时区，这个时候预选应该是 2024/09/30，并且点击此刻，生效的是 2024/09/30 22点10分。

如果上述没说错的话，应该修改 https://github.com/react-component/picker/blob/7689208b06098c4ceb98c7c69ca626a47dc7f930/src/generate/dayjs.ts#L110 的 `getNow` 就好了

```ts
 getNow: () => {
   return typeof (dayjs as any).tz === 'function'
      ? (dayjs as any).tz()
      : dayjs();
  },
```

欢迎讨论交流

**Reply by afc163** [reply]
问了下 GPT，确实可以这么操作。

---

在使用 `dayjs` 时，可以通过 `dayjs` 和 `dayjs.tz` （如果有安装 `dayjs-timezone` 插件）获取当前时区的当天日期。

如果没有特殊的时区处理，直接使用 `dayjs` 会返回基于当前系统时区的日期：

### 1. 不使用时区插件
```javascript
const dayjs = require('dayjs');

// 获取当前时区的当天日期（年月日）
const currentDate = dayjs().format('YYYY-MM-DD');
console.log(currentDate);
```
这会返回当前系统时区下的当天日期，格式为 `YYYY-MM-DD`。

### 2. 使用时区插件
如果你需要指定时区，可以安装 `dayjs-timezone` 插件。

首先，确保安装了 `dayjs-timezone` 插件：
```bash
npm install dayjs dayjs-plugin-timezone
```

然后，你可以使用以下代码获取特定时区的当天日期：

```javascript
const dayjs = require('dayjs');
const timezone = require('dayjs/plugin/timezone');
const utc = require('dayjs/plugin/utc');

dayjs.extend(utc);
dayjs.extend(timezone);

// 获取指定时区当天的日期，比如获取'Asia/Shanghai'时区
const currentDateInShanghai = dayjs().tz('Asia/Shanghai').format('YYYY-MM-DD');
console.log(currentDateInShanghai);
```

这样，你可以指定时区（如 `Asia/Shanghai`）来获取当天的日期。如果不指定时区，它将默认使用你当前的系统时区。

**Reply by yoyo837** [reply]
我系统用了dayjs的时区

**Reply by afc163** [reply]
重现链接：https://stackblitz.com/edit/react-s699lb?file=demo.tsx

```jsx
dayjs.tz.setDefault('America/Los_Angeles');
const now = dayjs().format('YYYY-MM-DD HH:mm:ss');
```

为什么下面这个 now 不是 Los_Angeles 的时间？

---

问题的原因在于，`dayjs().format('YYYY-MM-DD HH:mm:ss')` 这行代码没有明确指定时区。虽然你通过 `dayjs.tz.setDefault('America/Los_Angeles')` 设置了默认时区，但 `dayjs()` 仍然是使用本地时区进行计算。

要解决这个问题，你需要在获取当前时间时使用 `dayjs.tz()`，这样才能确保按照你设置的时区获取时间。

### 正确的做法
你应该显式地使用 `dayjs.tz()` 来获取指定时区的时间：

```javascript
const dayjs = require('dayjs');
const timezone = require('dayjs/plugin/timezone');
const utc = require('dayjs/plugin/utc');

dayjs.extend(utc);
dayjs.extend(timezone);

// 设置默认时区
dayjs.tz.setDefault('America/Los_Angeles');

// 使用 dayjs.tz() 显示默认时区的当前时间
const now = dayjs.tz().format('YYYY-MM-DD HH:mm:ss');
console.log(now); // 现在会是 Los Angeles 的时间
```

### 解释：
1. **`dayjs.tz.setDefault('America/Los_Angeles')`** 设置了默认时区为 `Los_Angeles`，但这只影响 `dayjs.tz()` 方法，未自动作用于普通的 `dayjs()` 实例。
2. **`dayjs.tz()`** 明确要求 `dayjs` 使用指定的时区或默认时区。
3. 你需要在获取当前时间时使用 `dayjs.tz()` 而不是 `dayjs()`，这样才能获得设定的时区时间。

**Reply by wowhaha7pupu** [reply]
这边先感谢各位大佬的回复🙏，但目前要解决 DatePicker 组件的预选实线蓝框(今天)的这个问题，好像还没解决

以上面的demo案例搭配使用者的情境来说：
1. 使用者当下的系统时间为：2024-09-21 14:09:01
2. 使用者于后台系统登入时，选定指定 "America/Los_Angeles" 的时区，此时洛杉矶时间：2024-09-20 23:09:01
3. 这时我们预设的立场是当使用者点击 DatePicker 组件的时候，预选的蓝色实线框框应该是要显示 20 号 而非 21 号，因为使用者已经切换成了 Los_Angeles 的时区

但目前好像没有提供任何api能够做到改变"预设蓝色实线框"这件事情，除非使用者去电脑设定里面更改系统时间到Los_Angeles的时区，才能使"预设蓝色实线框"选到20号。

**Reply by afc163** [reply]
需要按 @Wxh16144 给的方案改一下 rc-picker

**Reply by wowhaha7pupu** [reply]
看来的确只能这样了，感谢解答。👍

**Reply by afc163** [reply]
@wowhaha7pupu 直接来个 PR 如何？

**Reply by wowhaha7pupu** [reply]
抱歉怕能力还不够，刚入门不久，还在摸索中

**Comment by Wxh16144** [comment]
不确定 rc-picker 那边如何提交 PR 修复这个问题，只是针对 dayjs 的话倒是好改，但是其他的比如 momentjs dateFns 等日期库该起来还挺麻烦。

目前我能想到的方案有几个。

1. ~修改 node-modules 里面的 `rc-picker/lib/generate/dayjs` 用 npm-patch 的方式~ (不推荐
2. 重写 dayjsGenerateConfig 来覆盖 antd 的 DatePicker.

```tsx
import dayjsGenerateConfig from 'rc-picker/lib/generate/dayjs';
import dayjs from 'dayjs';

// dayjs plugin
import utc from 'dayjs/plugin/utc';
import timezone from 'dayjs/plugin/timezone';

dayjs.extend(utc);
dayjs.extend(timezone);


import { DatePicker as AntDatePicker } from 'antd';

const mergedGenerateConfig: typeof dayjsGenerateConfig = {
  ...dayjsGenerateConfig,
  getNow: () => {
    if (typeof dayjs.tz === 'function') {
      return dayjs.tz();
    }
    return dayjs();
  }
}

const originGeneratePicker = AntDatePicker.generatePicker;

const clonedAntDatePicker = originGeneratePicker(mergedGenerateConfig);

export default Object.assign({}, AntDatePicker, clonedAntDatePicker)

```

我特地写了一个 [Wxh16144/learn-timezone](https://github.com/Wxh16144/learn-timezone) 仓库来尝试时区操作。

上面的代码也可以在 https://github.com/Wxh16144/learn-timezone/tree/master/src/components/DatePicker 这里查看更多～

后面再看看 PR  怎么改进吧

**Reply by afc163** [reply]
就按你给的方案改吧：https://github.com/react-component/picker/blob/7689208b06098c4ceb98c7c69ca626a47dc7f930/src/generate/dayjs.ts#L110

**Comment by afc163** [comment]
https://github.com/react-component/picker/releases/tag/v4.6.15
